src/machine-c-api.h

```diff
@@ -82,1894 +83,528 @@ typedef enum { // NOLINT(modernize-use-using)
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-/// \details If the node is smaller than a page size, then it must lie entirely inside the same PMA range.
-CM_API int cm_get_proof(const cm_machine *m, uint64_t address, int log2_size, cm_merkle_tree_proof **proof,
-    char **err_msg);
-
-/// \brief  Deletes the instance of cm_merkle_tree_proof acquired from cm_get_proof
-/// \param proof Valid pointer to cm_merkle_tree_proof object
-CM_API void cm_delete_merkle_tree_proof(cm_merkle_tree_proof *proof);
-
-/// \brief Obtains the root hash of the Merkle tree
-/// \param m Pointer to valid machine instance
-/// \param hash Valid pointer to cm_hash structure that  receives the hash.
-/// \param err_msg Receives the error message if function execution fails
-/// or NULL in case of successful function execution. In case of failure error_msg
-/// must be deleted by the function caller using cm_delete_cstring.
-/// err_msg can be NULL, meaning the error message won't be received.
-/// \returns 0 for success, non zero code for error
-CM_API int cm_get_root_hash(const cm_machine *m, cm_hash *hash, char **err_msg);
+/// \param access_log Receives the state access log as a JSON string,
+/// remains valid until the next time this same function is called on the same thread.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_log_step_uarch(cm_machine *m, int log_type, bool one_based, const char **access_log);
+
+/// \brief Resets the entire uarch state to pristine values.
+/// \param m Pointer to a valid machine instance.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_reset_uarch(cm_machine *m);
+
+/// \brief Resets the value of the microarchitecture halt flag.
+/// \param m Pointer to a valid machine instance.
+/// \param log_type Type of access log to generate, must be bitwise OR of CM_ACCESS_LOG_TYPE enum.
+/// \param one_based Use 1-based indices when reporting errors.
+/// \param access_log Receives the state access log as a JSON string,
+/// remains valid until the next time this same function is called on the same thread.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_log_reset_uarch(cm_machine *m, int log_type, bool one_based, const char **access_log);
+
+/// \brief Sends cmio response.
+/// \param m Pointer to a valid machine instance.
+/// \param reason Reason for sending the response.
+/// \param data Response data to send
+/// \param length Length of response data.
+/// \returns 0 for success, non zero code for error.
+CM_API int cm_send_cmio_response(cm_machine *m, uint16_t reason, const unsigned char *data, size_t length);
```

Comment:
It would be nice to have the other way as well, something like this:
```
CM_API int cm_recv_cmio_request(cm_machine *m, uint16_t *reason,
    size_t max_length, unsigned char *data, size_t *length);
```

Reply:
I agree, I think this will be useful.

Reply:
We don’t implement this on the solidity side. So we don’t know what order they will need things in the log. We are also going to simplify this protocol soon, right?

Reply:
I see this as an abstraction of the HTIF + buffers interface we have today. The alternative is for users to pack / unpack HTIF by themselves, not ideal.

> We are also going to simplify this protocol soon, right?

I'm not following. Is this about uarch -> risc0 transition?
if so we could drop _log_ and _uarch_, but Node people will still need to load / retrieve bytes from buffers right?

Reply:
https://github.com/cartesi/machine-emulator/issues/257

Reply:
We can add a method for the reader node to use. But we won’t have a method for the verifier node to use. This would return a log. But we don’t control the blockchain implementation of whatever it is that would consume that log. So in what would be the information is it that we would provide ind the log and in what order?

Reply:
I think this could be useful on its own as a high level API to do IO with the machine through the C API for applications that want to use libcartesi as a C library. As of now after the machine yields you can send data inside with `cm_send_cmio_response` but there is no easy function to read data from inside, `cm_recv_cmio_request` would be this function.

Reply:
Go for it.

Reply:
> #257 

So not having to deal with `iflags`, only HTIF. That is how 

> So in what would be the information is it that we would provide ind the log and in what order?

Assuming: "So what would be the information we provide, and in what order?"

A bit out of my depth, but I don't see the issue. My understanding is that this is already a provable machine since it is the machine immediately before a surgery.

So for instance, we do `read_htif(...);` + `read_memory(TX_BUFFER);` in `cartesi-machine.lua`.
What is the difference from that to having a single function that does both and massages the internals?

Reply:
Because we don’t control what this code will look like in solidity. We don’t know anything about the different domains.

Reply:
The log has to include the places where the info comes from. So the changes impact this.